adafruit_onewire/bus.py

```diff
@@ -0,0 +1,206 @@
+
+    def __init__(self, rom):
+        self._rom = rom
+
+    @property
+    def rom(self):
+        """The unique 64 bit ROM code."""
+        return self._rom
+
+    @property
+    def crc(self):
+        """The 8 bit CRC."""
+        return self._rom[7]
+
+    @property
+    def serial_number(self):
+        """The 48 bit serial number."""
+        return self._rom[1:7]
+
+    @property
+    def family_code(self):
+        """The 8 bit family code."""
+        return self._rom[0]
+
+class OneWireBus(object):
+    """A class to represent a 1-Wire bus."""
+
+    def __init__(self, pin):
+        self._ow = busio.OneWire(pin)
+
+    def reset(self, required=False):
+        """
+        Perform a reset and check for presence pulse.
+
+        :param bool required: require presence pulse
+        """
+        reset = self._ow.reset()
+        if required and reset:
+            raise OneWireError
+        return not reset
+
+    def readinto(self, buf, **kwargs):
+        """
+        Read into ``buf`` from the device. The number of bytes read will be the
+        length of ``buf``.
+
+        If ``start`` or ``end`` is provided, then the buffer will be sliced
+        as if ``buf[start:end]``. This will not cause an allocation like
+        ``buf[start:end]`` will so it saves memory.
+
+        :param bytearray buf: buffer to write into
+        :param int start: Index to start writing at
+        :param int end: Index to write up to but not include
+        """
+        start = 0
+        end = len(buf)
+        if 'start' in kwargs:
+            start = kwargs['start']
+        if 'end' in kwargs:
+            end = kwargs['end']
+        for i in range(start, end):
+            buf[i] = self._readbyte()
+
+    def write(self, buf, **kwargs):
+        """
+        Write the bytes from ``buf`` to the device.
+
+        If ``start`` or ``end`` is provided, then the buffer will be sliced
+        as if ``buffer[start:end]``. This will not cause an allocation like
+        ``buffer[start:end]`` will so it saves memory.
+
+        :param bytearray buf: buffer containing the bytes to write
+        :param int start: Index to start writing from
+        :param int end: Index to read up to but not include
+        """
+        start = 0
+        end = len(buf)
+        if 'start' in kwargs:
+            start = kwargs['start']
+        if 'end' in kwargs:
+            end = kwargs['end']
+        for i in range(start, end):
+            self._writebyte(buf[i])
+
+    def scan(self):
+        """Scan for devices on the bus and return a list of addresses."""
+        devices = []
+        diff = 65
+        rom = False
+        for i in range(0xff): # pylint: disable=unused-variable
+            rom, diff = self._search_rom(rom, diff)
+            if rom:
+                devices += [OneWireAddress(rom)]
+            if diff == 0:
+                break
+        return devices
+
+    def _readbit(self):
+        return self._ow.read_bit()
```

Comment:
You could do `self._readbit = self._ow.read_bit` in the `__init__`, then it would be one function call less.

Reply:
i like doing it this way in case i ever want to come back and add additional behavior

Reply:
You can still do it anyways — you can switch to this way at any moment.

Reply:
is there any reason other than code style?

Reply:
It's actually a bit worse style-wise, but it's much faster -- function calls are expensive, and you are avoiding one this way, and also one frame on the stack. CircuitPython has pretty low recursion limit, it's easy to hit it with a lot of nesting.

Reply:
thanks. that's what i was after. :thumbsup: